File: VB/GridControlDetailView/CustomGrid/Registrator.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.XtraGrid.Registrator
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid
Imports DevExpress.Utils

Namespace CustomGrid
	Friend Class Registrator
		Inherits BaseInfoRegistrator
		Public Overrides ReadOnly Property ViewName() As String
			Get
				Return "CustomView"
			End Get
		End Property

		Public Overrides Function CreateView(ByVal grid As GridControl) As BaseView
			Return New CustomView(grid)
		End Function

		Public Overrides Function CreateViewInfo(ByVal view As BaseView) As DevExpress.XtraGrid.Views.Base.ViewInfo.BaseViewInfo
			Return New CustomViewInfo(TryCast(view, CustomView))
		End Function

		Public Overrides Function CreatePainter(ByVal view As BaseView) As BaseViewPainter
			Return New CustomPainter(TryCast(view, CustomView))
		End Function

		Public Overrides Function CreateHandler(ByVal view As BaseView) As DevExpress.XtraGrid.Views.Base.Handler.BaseViewHandler
			Return New CustomHandler(TryCast(view, CustomView))
		End Function

		Protected Overrides Sub RegisterViewPaintStyles()
			PaintStyles.Add(New FakeViewPaintStyle(Function() New CustomViewPaintStyle(), "Flat"))
		End Sub
	End Class

	Friend Delegate Function CreateViewPaintStyle() As ViewPaintStyle
	Friend Class FakeViewPaintStyle
		Inherits ViewPaintStyle
		Private creator As CreateViewPaintStyle
		Private name_Renamed As String

		Public Function Create() As ViewPaintStyle
			Return creator()
		End Function

		Public Sub New(ByVal creator As CreateViewPaintStyle, ByVal name As String)
			Me.creator = creator
			Me.name_Renamed = name
		End Sub

		Public Overrides ReadOnly Property Name() As String
			Get
				Return name_Renamed
			End Get
		End Property

		Public Overrides Function GetAppearanceDefaultInfo(ByVal view As BaseView) As AppearanceDefaultInfo()
			Throw New InvalidOperationException()
		End Function
	End Class

	Public Class CustomViewPaintStyle
		Inherits ViewPaintStyle
		Public Overrides Sub UpdateElementsLookAndFeel(ByVal view As BaseView)
			view.PaintStyleName = Name
		End Sub

		Public Overrides Function GetAppearanceDefaultInfo(ByVal view As BaseView) As AppearanceDefaultInfo()
			Throw New NotImplementedException()
		End Function

		Public Overrides ReadOnly Property Name() As String
			Get
				Return "Flat"
			End Get
		End Property
	End Class
End Namespace

File: CS/GridControlDetailView/CustomGrid/Registrator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DevExpress.XtraGrid.Registrator;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid;
using DevExpress.Utils;

namespace CustomGrid
{
    class Registrator : BaseInfoRegistrator
    {
        public override string ViewName
        {
            get
            {
                return "CustomView";
            }
        }

        public override BaseView CreateView(GridControl grid)
        {
            return new CustomView(grid);
        }

        public override DevExpress.XtraGrid.Views.Base.ViewInfo.BaseViewInfo CreateViewInfo(BaseView view)
        {
            return new CustomViewInfo(view as CustomView);
        }

        public override BaseViewPainter CreatePainter(BaseView view)
        {
            return new CustomPainter(view as CustomView);
        }

        public override DevExpress.XtraGrid.Views.Base.Handler.BaseViewHandler CreateHandler(BaseView view)
        {
            return new CustomHandler(view as CustomView);
        }

        protected override void RegisterViewPaintStyles()
        {
            PaintStyles.Add(new FakeViewPaintStyle(delegate
            {
                return new CustomViewPaintStyle();
            }, "Flat"));
        }
    }

    delegate ViewPaintStyle CreateViewPaintStyle();
    class FakeViewPaintStyle : ViewPaintStyle
    {
        CreateViewPaintStyle creator;
        string name;

        public ViewPaintStyle Create()
        {
            return creator();
        }

        public FakeViewPaintStyle(CreateViewPaintStyle creator, string name)
        {
            this.creator = creator;
            this.name = name;
        }

        public override string Name
        {
            get
            {
                return name;
            }
        }

        public override AppearanceDefaultInfo[] GetAppearanceDefaultInfo(BaseView view)
        {
            throw new InvalidOperationException();
        }
    }

    public class CustomViewPaintStyle : ViewPaintStyle
    {
        public override void UpdateElementsLookAndFeel(BaseView view)
        {
            view.PaintStyleName = Name;
        }

        public override AppearanceDefaultInfo[] GetAppearanceDefaultInfo(BaseView view)
        {
            throw new NotImplementedException();
        }

        public override string Name
        {
            get
            {
                return "Flat";
            }
        }
    }
}
